pkg/alertmanager/alertmanager.go

```diff
@@ -575,3 +607,117 @@ func (t *tenantRateLimits) RateLimit() rate.Limit {
 func (t *tenantRateLimits) Burst() int {
 	return t.limits.NotificationBurstSize(t.tenant, t.integration)
 }
+
+var (
+	errTooManyAlerts = "too many alerts, limit: %d"
+	errAlertsTooBig  = "alerts too big, total size limit: %d bytes"
+)
+
+const (
+	insertFailureTooManyAlerts = "too_many_alerts"
+	insertFailureAlertsTooBig  = "alerts_too_big"
+)
+
+type alertsLimiter struct {
+	tenant string
+	limits Limits
+
+	failureCounter *prometheus.CounterVec
+
+	mx        sync.Mutex
+	sizes     map[model.Fingerprint]int
+	count     int
+	totalSize int
+}
+
+func newAlertsLimiter(tenant string, limits Limits, failureCounter *prometheus.CounterVec) *alertsLimiter {
+	return &alertsLimiter{
+		tenant:         tenant,
+		limits:         limits,
+		sizes:          map[model.Fingerprint]int{},
+		failureCounter: failureCounter,
+	}
+}
+
+func (a *alertsLimiter) PreStore(alert *types.Alert, existing bool) error {
+	if alert == nil {
+		return nil
+	}
+
+	countLimit := a.limits.AlertmanagerMaxAlertsCount(a.tenant)
+	sizeLimit := a.limits.AlertmanagerMaxAlertsSizeBytes(a.tenant)
+
+	newSize := alertSize(alert.Alert)
+
+	a.mx.Lock()
+	defer a.mx.Unlock()
+
+	if !existing {
+		if countLimit > 0 && (a.count+1) > countLimit {
+			a.failureCounter.WithLabelValues(insertFailureTooManyAlerts).Inc()
+			return fmt.Errorf(errTooManyAlerts, countLimit)
+		}
+
+		if sizeLimit > 0 && (a.totalSize+newSize) > sizeLimit {
+			a.failureCounter.WithLabelValues(insertFailureAlertsTooBig).Inc()
+			return fmt.Errorf(errAlertsTooBig, sizeLimit)
+		}
+	}
+
+	return nil
+}
+
+func (a *alertsLimiter) PostStore(alert *types.Alert, existing bool) {
+	if alert == nil {
+		return
+	}
+
+	newSize := alertSize(alert.Alert)
+
+	a.mx.Lock()
+	defer a.mx.Unlock()
+
+	fp := alert.Fingerprint()
+	if existing {
+		a.totalSize -= a.sizes[fp]
+	} else {
+		a.count++
+	}
+	a.sizes[fp] = newSize
+	a.totalSize += newSize
+}
+
+func (a *alertsLimiter) PostDelete(alert *types.Alert) {
+	if alert == nil {
+		return
+	}
+
+	a.mx.Lock()
+	defer a.mx.Unlock()
+
+	fp := alert.Fingerprint()
+	a.totalSize -= a.sizes[fp]
+	delete(a.sizes, fp)
+	a.count--
+}
+
+func (a *alertsLimiter) currentStats() (count, totalSize int) {
+	a.mx.Lock()
+	defer a.mx.Unlock()
+
+	return a.count, a.totalSize
+}
+
+func alertSize(alert model.Alert) int {
+	size := 0
+	for l, v := range alert.Labels {
+		size += len(l)
+		size += len(v)
+	}
+	for l, v := range alert.Annotations {
+		size += len(l)
+		size += len(v)
+	}
+	size += len(alert.GeneratorURL)
```

Comment:
```suggestion
        // We include the generatorURL as part of the alert size given it includes the user-input query of the alert itself.
	size += len(alert.GeneratorURL)
```

Reply:
I'm mostly looking for a discussion point here (not saying it is wrong), I understand this is user input but a) we already have a limit for the size of the ruler configuration to catch egregiousness and b) it seems a bit unfair that we take a chunk of their limit based on something that might be beyond their control (they can reduce their query length but this also includes the URL of where it's going to)

Reply:
I think the alerts size limit (say 10 MiB) should be high enough that relatively short (< 50 chars) URL will not make a big difference.

I think including GeneratorURL makes sense because it better reflects size of the alert, if we do, but I see your point that user cannot control it, and I'd be fine removing it from the computation too.

Reply:
I do agree that it better reflects the size of the alert, let's leave it in. It's already documented as part of the flag which was my main concern.